src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs

```diff
@@ -283,11 +283,19 @@ public static void Move(string sourceDirName, string destDirName)
             if (!sourceRoot.Equals(destinationRoot, StringComparison.OrdinalIgnoreCase))
                 throw new IOException(SR.IO_SourceDestMustHaveSameRoot);
 
+            bool sourcePathIsFile = FileSystem.FileExists(fullsourceDirName);
```

Comment:
The problem of caching if the file exists, is that the file could exist initially, but then between this call and the subsequent places where the cached value is checked, the file could end up being deleted.